src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs

```diff
@@ -0,0 +1,170 @@
+                    if (!identity.HasClaim(c => c.Type == claim.Type))
+                    {
+                        identity.AddClaim(claim);
+                    }
+                }
+            }
+
+            // Get the SuperOffice user principal
+            using var request = new HttpRequestMessage(HttpMethod.Get, userInfoEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException($"An error occurred when retrieving SuperOffice user information ({response.StatusCode}). Please check if the authentication information is correct and the corresponding SuperOffice API is enabled.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+            var context = new OAuthCreatingTicketContext(new ClaimsPrincipal(identity), properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
+            if (Options.IncludeFunctionalRightsAsClaims
+                && !Options.ClaimActions.Any(c => c.ClaimType == SuperOfficeAuthenticationConstants.PrincipalNames.FunctionRights))
+            {
+                Options.ClaimActions.MapJsonKey(SuperOfficeAuthenticationConstants.PrincipalNames.FunctionRights, SuperOfficeAuthenticationConstants.PrincipalNames.FunctionRights);
+            }
+
+            context.RunClaimActions();
+            await Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        /// <summary>
+        /// Extracts the identity claims from the token received from the token endpoint.
+        /// </summary>
+        /// <param name="token">The token to extract the claims from.</param>
+        /// <returns>
+        /// An <see cref="IEnumerable{Claim}"/> containing the claims extracted from the token.
+        /// </returns>
+        protected virtual IEnumerable<Claim> ExtractClaimsFromToken([NotNull] string token)
+        {
+            try
+            {
+                var securityToken = _tokenHandler.ReadJwtToken(token);
```

Comment:
Basically, you're reading the `id_token` twice: once in the ID token validator and once here, which is not ideal. Let's centralize everything in the handler to avoid duplicate reading/validation.

Reply:
Moved all validation into the handler.